File: valdi/src/java/com/snap/valdi/promise/ResolvablePromise.kt
package com.snap.valdi.promise

import com.snap.valdi.utils.DisposableUtils
import java.util.concurrent.CancellationException

/**
 * Failure delivered to a canceled promise's callbacks when the producer does not settle it during
 * cancellation (see [Promise.cancel] contract in the C++ Promise.hpp). Mirrors the C++
 * kPromiseCanceledErrorCode failure so consumers can distinguish cancellation from a genuine error.
 *
 * Being a [CancellationException] is what carries the code across the JNI boundary:
 * `Throwable.valdiErrorCode()` maps it to `PROMISE_CANCELED_ERROR_CODE`, which
 * `CppPromiseCallback.onFailure` passes to native alongside the message. In the other direction a
 * coded native failure arrives as a `ValdiException` with `errorCode` set.
 */
class PromiseCanceledException : CancellationException("Promise canceled")

open class ResolvablePromise<T>: Promise<T> {

    private var completions: MutableList<PromiseCallback<T>>? = null
    private var completed = false
    private var error: Throwable? = null
    private var value: T? = null
    private var canceled = false

    fun fulfillSuccess(value: T) {
        var completions: MutableList<PromiseCallback<T>>?
        synchronized(this) {
            // A producer may legitimately settle after cancel — its synchronous settle from
            // onCancel is what forwards the real result; only a completed promise drops the result.
            if (completed) {
                return
            }
            completed = true
            this.value = value
            completions = this.completions
            this.completions = null
        }

        completions?.forEach {
            it.onSuccess(value)
            DisposableUtils.disposeAny(it)
        }
    }

    fun fulfillFailure(error: Throwable) {
        var completions: MutableList<PromiseCallback<T>>?
        synchronized(this) {
            if (completed) {
                return
            }
            completed = true
            this.error = error
            completions = this.completions
            this.completions = null
        }

        completions?.forEach {
            it.onFailure(error)
            DisposableUtils.disposeAny(it)
        }
    }

    override fun onComplete(callback: PromiseCallback<T>) {
        var error: Throwable?
        var value: T?
        synchronized(this) {
            // A registrant that lands between doCancel and finishCancel is appended and drained
            // by finishCancel or by a synchronous producer settle.
            if (!completed) {
                var completions = this.completions
                if (completions == null) {
                    completions = mutableListOf()
                    this.completions = completions
                }
                completions.add(callback)
                return
            } else {
                error = this.error
                value = this.value
                // We are already completed
                // Unlock the mutex and call the completion with the result
            }
        }

        if (error != null) {
            callback.onFailure(error!!)
        } else {
            callback.onSuccess(value!!)
        }
        DisposableUtils.disposeAny(callback)
    }

    override fun cancel() {
        if (doCancel()) {
            finishCancel()
        }
    }

    override fun isCancelable(): Boolean {
        return false
    }

    protected fun doCancel(): Boolean {
        synchronized(this) {
            if (canceled || completed) {
                return false
            }
            this.canceled = true
        }
        return true
    }

    /**
     * Second phase of cancellation, run after the producer's cancel hook had a chance to settle the
     * promise: if it did, the callbacks already received the real result; otherwise settle with
     * [PromiseCanceledException] and forward it to (and release) the pending callbacks.
     */
    protected fun finishCancel() {
        var completions: MutableList<PromiseCallback<T>>?
        val error: Throwable
        synchronized(this) {
            if (completed) {
                return
            }
            error = PromiseCanceledException()
            this.error = error
            completed = true
            completions = this.completions
            this.completions = null
        }

        completions?.forEach {
            it.onFailure(error)
            DisposableUtils.disposeAny(it)
        }
    }

}

File: valdi_core/src/valdi_core/cpp/Utils/ResolvablePromise.cpp
//
//  Error.cpp
//  valdi-ios
//
//  Created by Simon Corsin on 8/30/19.
//

#include "valdi_core/cpp/Utils/ResolvablePromise.hpp"
#include "utils/debugging/Assert.hpp"
#include "valdi_core/cpp/Utils/StringCache.hpp"

namespace Valdi {

namespace {

// Interned once: STRING_LITERAL hashes and takes the global StringCache mutex on every call, and
// cancel() runs on teardown paths.
STRING_CONST(promiseCanceledMessage, "Promise canceled")

} // namespace

ResolvablePromise::ResolvablePromise() = default;
ResolvablePromise::~ResolvablePromise() = default;

void ResolvablePromise::fulfillWithPromiseResult(const Ref<Promise>& promise) {
    promise->onComplete([self = strongSmallRef(this)](const auto& result) mutable {
        if (self != nullptr) {
            self->fulfill(result);
            self = {};
        }
    });

    if (promise->isCancelable()) {
        setCancelCallback([promise]() {
            // This captures a strong ref reference, which will form a cycle.
            // When this promise is fulfilled by upstream promise,
            // the callback is removed, and the cycle will break
            if (promise != nullptr) {
                promise->cancel();
            }
        });
    }
}

void ResolvablePromise::fulfill(Result<Value> result) {
    std::unique_lock<Mutex> lock(_mutex);
    if (!_result.empty()) {
        // A producer may legitimately settle asynchronously after cancel() already recorded the
        // synthetic canceled error; drop its result. Settling twice without a cancel is still a bug.
        SC_ASSERT(_canceled, _result.description());
        return;
    }

    _result = result;
    // clear after move because moved-from function is in "valid but unspecified" state
    auto callbacks = std::move(_callbacks);
    _callbacks = {};
    auto cancelCallback = std::move(_cancelCallback);
    _cancelCallback = {};

    lock.unlock();

    for (const auto& callback : callbacks) {
        if (result) {
            callback->onSuccess(result.value());
        } else {
            callback->onFailure(result.error());
        }
    }
}

void ResolvablePromise::onComplete(const Ref<PromiseCallback>& callback) {
    std::unique_lock<Mutex> lock(_mutex);
    // A registrant that lands between cancel()'s two phases (canceled, result not recorded yet) is
    // appended and drained by cancel()'s second phase or by a synchronous producer settle.
    if (!_result.empty()) {
        auto result = _result;
        lock.unlock();

        if (result) {
            callback->onSuccess(result.value());
        } else {
            callback->onFailure(result.error());
        }

        return;
    }

    _callbacks.emplace_back(callback);
}

bool ResolvablePromise::isCancelable() const {
    std::lock_guard<Mutex> lock(_mutex);
    return _cancelCallback;
}

void ResolvablePromise::cancel() {
    std::unique_lock<Mutex> lock(_mutex);
    if (_canceled || !_result.empty()) {
        return;
    }
    _canceled = true;
    auto cancelCallback = std::move(_cancelCallback);
    _cancelCallback = {}; // clear after move

    // _callbacks intentionally stays populated across the cancel callback: the producer may settle
    // synchronously from it, and fulfill() then drains the callbacks with the real result.
    lock.unlock();

    if (cancelCallback) {
        cancelCallback();
    }

    lock.lock();
    if (!_result.empty()) {
        // The producer settled during cancellation; the consumers already got the real result.
        // Unlock before returning so cancelCallback — which can hold the last ref to the upstream
        // promise, and on iOS bottom out in an ObjC -dealloc — is destroyed with the mutex released.
        lock.unlock();
        return;
    }
    _result = Error(promiseCanceledMessage(), kPromiseCanceledErrorCode);
    auto canceledError = _result.error();
    auto callbacks = std::move(_callbacks);
    _callbacks = {}; // clear after move
    lock.unlock();

    for (const auto& callback : callbacks) {
        callback->onFailure(canceledError);
    }
}

void ResolvablePromise::setCancelCallback(DispatchFunction cancelCallback) {
    std::unique_lock<Mutex> lock(_mutex);
    // Check _canceled before _result: cancel() records a result, and a producer attaching to an
    // already-canceled promise must still be told to cancel rather than silently dropped.
    if (_canceled) {
        lock.unlock();

        if (cancelCallback) {
            cancelCallback();
        }
        return;
    }

    if (!_result.empty()) {
        lock.unlock();
        return;
    }

    auto oldCancelCallback = std::move(_cancelCallback);
    _cancelCallback = std::move(cancelCallback);
    lock.unlock();
}

} // namespace Valdi
